File: copilot-token-budget/extension/src/livebilling/config.test.ts
import * as assert from 'assert';
import * as fs from 'fs';
import * as os from 'os';
import * as path from 'path';
import { defaults, loadLiveBillingConfig, resolveLiveBillingAuth } from './config';

class TestSuite {
  private tests: Array<{ name: string; fn: () => void | Promise<void> }> = [];

  test(name: string, fn: () => void | Promise<void>): void {
    this.tests.push({ name, fn });
  }

  async run(): Promise<void> {
    let failed = 0;
    for (const test of this.tests) {
      try {
        await test.fn();
        console.log(`✓ ${test.name}`);
      } catch (err) {
        failed += 1;
        console.error(`✗ ${test.name}: ${err}`);
      }
    }
    if (failed > 0) {
      process.exit(1);
    }
  }
}

const suite = new TestSuite();

suite.test('defaults are opt-in and token-env based', () => {
  const cfg = defaults();
  assert.strictEqual(cfg.enabled, false);
  assert.strictEqual(cfg.tokenEnvVar, 'COPILOT_BILLING_TOKEN');
  assert.strictEqual(cfg.cacheMaxAgeHours, 24);
  assert.strictEqual(cfg.requestTimeoutSecs, 10);
});

suite.test('loadLiveBillingConfig merges config.json over defaults', () => {
  const root = fs.mkdtempSync(path.join(os.tmpdir(), 'copilot-livebilling-'));
  const originalHome = process.env.HOME;
  const originalAppData = process.env.APPDATA;
  process.env.HOME = root;
  process.env.APPDATA = root;

  try {
    const configDir = process.platform === 'win32'
      ? path.join(root, 'copilot-token-budget')
      : path.join(root, '.config', 'copilot-token-budget');
    fs.mkdirSync(configDir, { recursive: true });
    fs.writeFileSync(
      path.join(configDir, 'config.json'),
      JSON.stringify({
        enabled: true,
        orgSlug: 'att-enterprise',
        tokenEnvVar: 'BILLING_TOKEN',
        cacheMaxAgeHours: 48,
        requestTimeoutSecs: 20,
        dryRun: true,
      }),
    );

    const cfg = loadLiveBillingConfig();
    assert.strictEqual(cfg.enabled, true);
    assert.strictEqual(cfg.orgSlug, 'att-enterprise');
    assert.strictEqual(cfg.tokenEnvVar, 'BILLING_TOKEN');
    assert.strictEqual(cfg.cacheMaxAgeHours, 48);
    assert.strictEqual(cfg.requestTimeoutSecs, 20);
    assert.strictEqual(cfg.dryRun, true);
  } finally {
    if (originalHome === undefined) {
      delete process.env.HOME;
    } else {
      process.env.HOME = originalHome;
    }
    if (originalAppData === undefined) {
      delete process.env.APPDATA;
    } else {
      process.env.APPDATA = originalAppData;
    }
  }
});

suite.test('resolveLiveBillingAuth reports disabled, dry-run, and ready modes', () => {
  const disabled = resolveLiveBillingAuth(defaults());
  assert.strictEqual(disabled.mode, 'disabled');
  assert.strictEqual(disabled.ready, false);
  assert.strictEqual(disabled.disabled, true);

  const dryRun = resolveLiveBillingAuth({
    enabled: true,
    orgSlug: 'att-enterprise',
    tokenEnvVar: 'COPILOT_BILLING_TOKEN',
    cacheMaxAgeHours: 24,
    requestTimeoutSecs: 10,
    dryRun: true,
  });
  assert.strictEqual(dryRun.mode, 'dry-run');
  assert.strictEqual(dryRun.ready, false);

  const ready = resolveLiveBillingAuth({
    enabled: true,
    orgSlug: 'att-enterprise',
    tokenEnvVar: 'COPILOT_BILLING_TOKEN',
    cacheMaxAgeHours: 24,
    requestTimeoutSecs: 10,
    dryRun: false,
  }, {
    COPILOT_BILLING_TOKEN: 'secret-token',
  });
  assert.strictEqual(ready.mode, 'ready');
  assert.strictEqual(ready.ready, true);
  assert.strictEqual(ready.token, 'secret-token');
});

void suite.run();

File: copilot-token-budget/core/internal/livebilling/config_test.go
package livebilling

import (
	"encoding/json"
	"os"
	"path/filepath"
	"testing"

	"github.com/aaraminds/copilot-token-budget/internal/platform"
)

func useTempConfig(t *testing.T) string {
	t.Helper()
	tmp := t.TempDir()
	t.Setenv("XDG_CONFIG_HOME", tmp)
	t.Setenv("HOME", tmp)
	t.Setenv("AppData", tmp)
	dir, err := platform.ConfigDir()
	if err != nil {
		t.Fatalf("ConfigDir: %v", err)
	}
	return dir
}

func TestDefault(t *testing.T) {
	c := Default()
	if c.Enabled {
		t.Fatal("Enabled = true, want false")
	}
	if c.TokenEnvVar != defaultTokenEnvVar {
		t.Fatalf("TokenEnvVar = %q, want %q", c.TokenEnvVar, defaultTokenEnvVar)
	}
	if c.CacheMaxAgeHours != defaultCacheMaxAge {
		t.Fatalf("CacheMaxAgeHours = %d, want %d", c.CacheMaxAgeHours, defaultCacheMaxAge)
	}
	if c.RequestTimeoutSecs != defaultRequestSecs {
		t.Fatalf("RequestTimeoutSecs = %d, want %d", c.RequestTimeoutSecs, defaultRequestSecs)
	}
}

func TestLoad_MergesOverDefaults(t *testing.T) {
	dir := useTempConfig(t)
	if err := os.MkdirAll(dir, 0700); err != nil {
		t.Fatal(err)
	}

	override := Config{
		Enabled:            true,
		OrgSlug:            "att-enterprise",
		TokenEnvVar:        "BILLING_TOKEN",
		CacheMaxAgeHours:   48,
		RequestTimeoutSecs: 20,
		DryRun:             true,
	}
	data, _ := json.Marshal(override)
	if err := os.WriteFile(filepath.Join(dir, configFileName), data, 0600); err != nil {
		t.Fatal(err)
	}

	c, err := Load()
	if err != nil {
		t.Fatalf("Load: %v", err)
	}
	if !c.Enabled || c.OrgSlug != "att-enterprise" {
		t.Fatalf("Load() = %+v, want merged override", c)
	}
	if c.TokenEnvVar != "BILLING_TOKEN" || c.CacheMaxAgeHours != 48 || c.RequestTimeoutSecs != 20 || !c.DryRun {
		t.Fatalf("Load() = %+v, want override values", c)
	}
}

func TestResolveAuth_Modes(t *testing.T) {
	disabled := ResolveAuth(Default(), os.LookupEnv)
	if disabled.Mode != "disabled" || !disabled.Disabled || disabled.Ready {
		t.Fatalf("disabled resolution = %+v", disabled)
	}

	missingOrg := ResolveAuth(Config{Enabled: true, TokenEnvVar: defaultTokenEnvVar, CacheMaxAgeHours: 24, RequestTimeoutSecs: 10}, os.LookupEnv)
	if missingOrg.Mode != "config-error" || missingOrg.Ready {
		t.Fatalf("missing org resolution = %+v", missingOrg)
	}

	dryRun := ResolveAuth(Config{
		Enabled:            true,
		OrgSlug:            "att-enterprise",
		TokenEnvVar:        defaultTokenEnvVar,
		CacheMaxAgeHours:   24,
		RequestTimeoutSecs: 10,
		DryRun:             true,
	}, os.LookupEnv)
	if dryRun.Mode != "dry-run" || dryRun.Ready || dryRun.HasToken {
		t.Fatalf("dry-run resolution = %+v", dryRun)
	}

	t.Setenv(defaultTokenEnvVar, "secret-token")
	ready := ResolveAuth(Config{
		Enabled:            true,
		OrgSlug:            "att-enterprise",
		TokenEnvVar:        defaultTokenEnvVar,
		CacheMaxAgeHours:   24,
		RequestTimeoutSecs: 10,
	}, os.LookupEnv)
	if ready.Mode != "ready" || !ready.Ready || !ready.HasToken || ready.Token != "secret-token" {
		t.Fatalf("ready resolution = %+v", ready)
	}
}
